logstash-core/src/main/java/org/logstash/common/io/MemoryCheckpointIO.java

```diff
@@ -12,7 +15,7 @@
     private final String HEAD_CHECKPOINT = "checkpoint.head";
     private final String TAIL_CHECKPOINT = "checkpoint.";
 
-    private static final Map<String, Checkpoint> sources = new HashMap<>();
+    private static final Map<String, Map<String, Checkpoint>> sources = new HashMap<>();
```

Comment:
all classes below the `Queue` class are not threadsafe, but all Queue operations are wrapped in a mutex so the Checkpoint operations are safe in that respect.